### DISCUSSION TITLE
add_subdomain_ids and Exodus output

### DISCUSSION THREAD

**Discussion by Wendy-Ji** [idaholab/moose#28485] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hello, 

I am using `add_subdomain_ids` to add an empty subdomain (see below code). I need this for my simulations which start off with empty subdomains and have elements later added to them. 

When I turn on the Exodus output, I get the below error. If the variable isn't `MONOMIAL` and `CONSTANT`, then this error won't appear. The same problem occurs if I use `add_subdomain_names`. I'm not sure if I did something wrong with the setup, or if you can't have an Exodus output for this particular setup.

Thank you.

Wendy

![image](https://github.com/user-attachments/assets/d24f9a6f-2309-49e3-b8e4-03f3e9460364)

```
[Problem]
  solve = false
[]

[Mesh]
  [generated]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 10
    ny = 10
  []
  add_subdomain_ids = 1
[]

[Variables]
  [dummy]
    family = MONOMIAL
    order = CONSTANT
    block = '0 1'
  []
[]

[Preconditioning]
  [SMP]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient
  end_time = 2
[]

[Outputs]
  exodus = True
[]
```

**Comment by GiudGiud** [comment]
hello

this is a known issue, you need to actually have an element in subdomain 1 before outputting to exodus (for now)

Guillaume

**Reply by Wendy-Ji** [reply]
Oh, I guess I haven't run into the problem before because I've been using the default family/order for variables, and that outputs to exodus fine. Is it going to be fixed?

**Reply by jmeier** [reply]
Dear Guillaume, I have the same problem right now (`ExodusII_IO_Helper: block id XXX not found in block_ids.`). It makes it difficult for us to use the [TimedSubdomainModifier ](https://mooseframework.inl.gov/docs/site/source/meshmodifiers/TimedSubdomainModifier.html).

You wrote there must be at least one element in each subdomain before outputting to exodus. In each timstep? Or just the first one?

The error message states the error is thrown in libmesh. According to a backtrace I did with gdb Moose is calling libmesh in line 343 of Exodus.C (method `Exodus::outputElementalVariables()`). Could we insert some kind of if-condition to avoid the error? What should be checked?
https://github.com/idaholab/moose/blob/82b674a17dca177b6c140de82e9f4ceb14f097ad/framework/src/outputs/Exodus.C#L333-L343

**Reply by GiudGiud** [reply]
We have no investigated it further. 

We could remove the non-existing block from the element data before sending it to the exodus writer. That would be fine imo but it s a little hacky. The better fix is in libmesh, modifying write_element_data to not die on missing subdomains

**Reply by jmeier** [reply]
Should I create an issue [over there](https://github.com/libMesh/libmesh/issues)? Or do you have better connections with the guys there?

**Reply by GiudGiud** [reply]
Sure

**Reply by jmeier** [reply]
Please find the issue with libmesh here: https://github.com/libMesh/libmesh/issues/3949
